### DISCUSSION TITLE
Computing residual using an input vector

### DISCUSSION THREAD

**Discussion by charlieowen3065** [idaholab/moose#26447] [Q&A General]
Hello and Happy New Year!

In MOOSE, I am generating a new solution vector for temperature for a heat conduction problem using a predictor. The predictor takes in the initial guess at the beginning of each time step then predicts a new solution and uses that as the initial guess. 

To see if my predictor is working well, I would ideally like to calculate the residual using both the original initial guess for the timestep, along with my new vector, so I would be calculating the residual using two different vectors.

My question is if there is any functionality in MOOSE that already exists to do this, or would I have to take an approach similar to that in ArrayDiffsusion.C (https://mooseframework.inl.gov/source/variables/ArrayMooseVariable.html) where they wrote in the governing equation and solved there.

That would be fine to do, but I was just wondering if you all had an internal method to do this already made or if there was another good example to look at for how to do this.

Thanks you!

Charlie

**Comment by GiudGiud** [comment]
@lindsayad @grmnptr @oanaoana either of you used an API that does this before?

**Comment by grmnptr** [comment]
Look for `FEProblemBase::computeResidual` and `NonlinearSystemBase::computeResidual` functions in the doxygen. Most of them take in a solution vector which is used to evaluate the residual at.

**Reply by charlieowen3065** [reply]
Yep, found those! Thanks!

I am running into a little issue with them, but I feel like it may just be a C++ issue. 

What is wrong with this format?

```
NonlinearSystemBase & nonlinear_sys = FEProblemBase::getNonlinearSystemBase(/*nl_sys=*/0);

NumericVector<Real> & residual_org;
TagID tag_org = 0;

nonlinear_sys.computeResidualTag(residual_org, tag_org);
```

I am getting hit with an error for `NonlinearSystemBase & nonlinear_sys = FEProblemBase::getNonlinearSystemBase(/*nl_sys=*/0)` 
(`error: cannot call member function 'NonlinearSystemBase& FEProblemBase::getNonlinearSystemBase(unsigned int)' without object
  210 |  NonlinearSystemBase & nonlinear_sys = FEProblemBase::getNonlinearSystemBase(/*nl_sys=*/0);`) 
and 
`NumericVector<Real> & residual_org;` 
(`error: 'residual_org' declared as reference but not initialized
  213 |  NumericVector<Real> & residual_org;`)

I tried to inherit FEProblemBase, but that made `getParams(...)` ambigious, so that didn't seem right, so I only include `FEProblemBase.h`, not inherit from it.

I pulled some of the basic ideas from `ReferenceResidualProblem.C`

Thanks!

**Reply by GiudGiud** [reply]
The error message is giving you a clue. You cannot use a reference for the residual_org vector because you are not initializing it. Either initialize it or create it by value instead of a reference (just delete the &)
